### ISSUE TITLE
sharing of complete password vaults

### ISSUE THREAD

**Issue by budachst** [nextcloud/passman#243]
In lastpass I am using the shared-password vault to setup some kind of family vault, where I can share and manage passwords that are available to other family members. This would make a great feature for passman to offer.

**Comment by cowai** [comment]
This would also benefit teams where there shouldnt be a single owner.

**Comment by cmcotte** [comment]
i would love to share vaults with team members. so i would like to choose to share with users and with groups. i really need this feature :(

**Comment by animalillo** [comment]
well, right now implementing this feature in a secure way is out of our current priorities, you are welcome to create a fork and make a pull request, we will review your code, and if it's as secure as our current implementation of sharing, we will merge it.

Right now we are busy making plugins and apps for passman, you know, we can't cover everything, we have a life, and we also need food and sleep!

**Comment by budachst** [comment]
Sure, I fully understand your take on this. If I had to choose, I'd go with making plugins first as well. However, this still would make a very nice and useful feature.

**Comment by Ninos** [comment]
Hey there, I would also love to see such functionality. ATM I cannot use your password manager because such feature is missing (our team needs access to a customer container, we cannot share each password to each user).
If a donation speed up the implementation of this feature, just let me know. I'm sure some guys would also love to donate for such a feature :-) I'm also ok if you want to sell this feature in first time (e.g. after reaching 100 sells -> implementing into core).

**Comment by metalcated** [comment]
Yeah I would also be willing to donate some coins. I am also having the same hurdle, would be much easied to share an entire vault. Thanks!

**Comment by brantje** [comment]
**Feature request accepted**
Implementing this feature is gonna take time. We ( @animalillo and me), are willing to take a few day's (3 - 4) off from our full time jobs to implement this feature. Therefore donations to cover those day's would be gladly accepted. I made a [special](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=A6DEAUFXALU4W) donation link for this feature, so we can track the donations for this feature.

**Comment by Ninos** [comment]
I've donated 50 USD, I'll donate 50 more after implementation. Thank you guys!

**Comment by brantje** [comment]
Thank you very much @Ninos it's much appreciated

**Comment by budachst** [comment]
Let me see, if I can get my employer to let me spend some money on this as well. Also, my dutch really is not that good. ;) I'd really like to get to the donation page in english…

**Comment by brantje** [comment]
@budachst I think it's fixed.

**Comment by budachst** [comment]
Yeah - great! Thanks… now let me try to collect some money… ;)

**Comment by brantje** [comment]
Thanks Mr F.W (using initials for privacy), for your generous donation ($150).
Please comment if you want to use your github username in the donator list.

**Comment by budachst** [comment]
Hey, that was my company. :)

**Comment by brantje** [comment]
@budachst Thank you very much. How would you like to be listed?

**Comment by budachst** [comment]
Actually, our company name would be nice, if that'd be possible: Jung von Matt AG

**Comment by brantje** [comment]
Everyone is welcome in our new telegram channel: https://t.me/passman_general
Purpose: general talk, off topic chat

**Comment by gschoenberger** [comment]
This would be a great feature! Currently I am search for a password manager for teams (self hosted, open source) and I must admit passman is currently by far the best looking option.

**Comment by Ninos** [comment]
@brantje is it possible to use slack or something like that? I don't have telegram.

BTW: For the container please use the users login password (user private key, decrypted via nextcloud with the user login password). Otherwise sharing containers with others will be very hard, because then you also need to send them the container password (for each container).

**Comment by animalillo** [comment]
We will implement it @gschoenberger, we don't have an ETA yet, but it's on the priorities list now, so expect it in the near future. We sadly can't do magic, tho, I wish we could :D

**Comment by Dark-Schnitzel** [comment]
Will it also be able to share Certain Tags with other Vaults?
Great Job you are doing so far!

**Comment by gschoenberger** [comment]
Concerning password sharing - which is IMO a killer feature for teams - is it currently possible to en-/decrypt a single vault with multiple passwords?
E.g. a team has a single vault and one single shared user to access the vault but each user knows a different password to decrypt the vault?
That would be a simple solution to share a vault without using only one password, like it is done e.g. with cryptsetup key slots (dm-crypt).

**Comment by animalillo** [comment]
We are going to dive into this issue this weekend and a few days next week, we will keep you updated!

**Comment by xrkolovos** [comment]
Thank you. I talk with my company, and we are going to donated too, for this feature.!

**Comment by brantje** [comment]
Just letting you guy's know that we didn't forget this feature.
At the moment the feature is about 30% ready, it is taking more time that we initially expected.

**Comment by pafcioooo** [comment]
Can anybody describe how exactly this feature will work? It with be great if the description would refer to what is in #340. Thanks.

**Comment by animalillo** [comment]
automatically adding people in groups access to shared password is not going to be implemented anytime soon due the way sharing works, everything is encrypted client side, so we have no way to share the password with a group currently

As of how will this work, it will be very similar to a normal share, except it will be performed from the vault settings page.

We have yet to decide if we include the share as a link inside a vault or ask the user to enter a vault key upon share completion to encrypt the shared vault shared key and then show the vault as another vault

**Comment by Ninos** [comment]
Why you're not using the user/group keys as NextCloud is doing it? It's much easier for you, you only need to encrypt/decrypt with the NC keys. Other things are done by NC.

**Comment by animalillo** [comment]
The way passman handles sharing and encryption is way diffrent from nextcloud, it's full clientside encryption, that's the problem, the user key never ever touches the server, so to share it we must fulfill that to keep the current passman secutiry

**Comment by Ninos** [comment]
Thanks for the answer. Normally the user keys are also secured with the user login password in NC, as I know correctly.

**Comment by animalillo** [comment]
The encryption owncloud/nextcloud performs is based on the user's login password, but it's performed on the server side, which makes auto sharing way easier.

The way we handle our encryption is on the client side, so the user key never touches the server. Then we encrypt it again on the server so in case the database gets leaked it's even harder to get to decrypt the vaults (get the server encryption key + clientside encryption key).

The process to view a passman item requires:
- Server decryption with server side key
- Client decryption with client side key

So if you access the vault trough the server it does the decryption normally, but then it sends you a bunch of encrypted data that has to be decrypted on your browser.

**Comment by Ninos** [comment]
OK understood, thx. But for me it's ok if the passwords get decrypted server-side by user password (linke OC/NC is handle it). If the server gets compromitted, you have much other problems than that. The hacker can also send manipulated js-files :-) I think server-side is much enough secure and also much better in ux. Think about big companies. The people should just have one password for the complete NC instance, otherwise they'll select not strong enough passwords :)

**Comment by maestroi** [comment]
So for example if i own the server where you hosting your passwords you have no problem that i can read them?

**Comment by animalillo** [comment]
@Ninos you can always use our browser extension or android application which can't be hacked server side! ;·P

Or use it as an API so we don't have to trust server provided js!

It's on passman philosophy to keep everything client side encrypted, if this complicates some stuff in exange of security, that's something we don't mind.

For us, passwords are a precious resource!

**Comment by pafcioooo** [comment]
I agree with @Ninos. Also smaller companies, like we are, want to have both security and easiness (it always need to be balanced). With #340 we currently stay before sharing problem though we have only 5 people. We have like 500 passwords and we can't move if the sharing would be vault based: this is why we did not go with lastpass.

I am looking at this from another side. If I am using some public service than I would go with client side encryption to be sure no one is reading my passwords. If I go with my company own hosting password server, then those are company passwords anyway. Of course in this later case, it should be additionally protected so reading would not be so easy. 

I am not a coder, but if there is possibility to make users life both secure and easy, then maybe it is better to make a few steps back, and think if there is some better way.
Thanks.

**Comment by brantje** [comment]
We won't drop client side encryption ever, that is one of the features of Passman.

**Comment by pafcioooo** [comment]
@brantje cool:-) Please make sharing easier then:-) So we can share single password or group of passwords both for a single user and a group of users:-)

**Comment by animalillo** [comment]
we are adding this feature so you can share a whole vault, if that's not sharing a gorup of passwords i don't know what is sharing a group of things anymore o:

**Comment by maestroi** [comment]
@pafcioooo  so you are suggesting this password manager should only be for small teams? 
also shareing passwords in most situations are bad, for example if someone used your credentials for some critical sevice you are the one to blame because you shared your credentials.

we know some people want sharing and we are looking into it but its far less secure than not sharing.

**Comment by pafcioooo** [comment]
I am saying that sharing passwords using passman, even in small team, is not easy. We have tried it and decided to stay with WebPasswordSafe, old java app, which is not so secure but makes sharing those 500 passwords, much easier - but this is server side encryption.
This is why I am suggesting that sharing with passman within teams should be easier. In my opinion, with current vault idea, vault sharing, with each vault having its own password (which with web browser extension requires changing vault that to you currently read/write), would be more than cumbersome. 
Security is a top concern for me. But another top concern is user experience.

**Comment by brantje** [comment]
`(which with web browser extension requires changing vault that to you currently read/write)`
Not really, version 2.0 of the extension adds the possibility of having multiple accounts / vaults installed.
The extension will use all those sources to find matching accounts.

**Comment by pafcioooo** [comment]
@Brantje thanks for that:-)

**Comment by Ninos** [comment]
Sorry for the late response, I was in holidays. Can you tell me the exact process for password sharing. 
- Do my team and I need additional passwords for the vaults or just the users login password once time?
- Is it possible for a new group member to access the vaults without additional tasks of the vaults owner?

Please please think about small, middle and big companies. For us it's important to have the passwords encrypted in database, but it's enough if they can be decrypted users password (private key...).
1. For companies it's more secure to have a strong login password for all vaults instead of having an own password for each vault (leads them to write the passwords somewhere else [may under their keyboard?].
2. For companies it's important to be able to add/remove members from groups without additional tasks. So the new members should be able to access the vaults without additional tasks.

I hope it will be so, otherwise my donation was for nothing and I could also use keepass. And about the client-side encrpytion, please don't argue that I can use the android app or browser extension. NextCloud is thought to be a collaboration software, so using NextCloud via browser should be first prio and here it's also possible to send manipulated js-files, if server is compromitted. Think about the future. In future companies want to handle all of their tasks online through a collaboration software like NextCloud.

**Comment by animalillo** [comment]
@Ninos what I can tell you so far is that the shared vaults would be integrated in a way that the target users either:
  a) get them as an special item inside their vault that opens a new vault when clicked
  b) have to type a password to access that vault upon share completion
  c) keep both options and let the user chose upon vault share acceptance

As for the sharing process it would be pretty much like the current credential item share process.

I can't assure you which path we will chose at this point, but it will depend on the challenges implementing them.

As for the automatic addition of users into shares, nextcloud user creation and edition hooks are not compatible with our sharing system. You must understand that passman is a full app on itself and that it is integrated within the owncloud/nextcloud ecosystem because we think it is easier for people to get everything in the same site.

As of this last point, we are not going to change that anytime soon, we might, however add at some point bulk actions so you can apply a share to a whole group, or change permissions to a whole group. But that might come in the future, we are an small team and we do this in our spare time for the most part, altought we have booked some vacation days to work on sharing, we can't quit our jobs and work full time on passman doing everything we would love this app to have. And believe us, we would love to be able to work full time on passman and get all this small things running.

What we are not going to do is lower our security standards, but if you feel like some feature is very easy to do, we would gladly review any merge requests that come as soon as we can.

**Comment by creolis** [comment]
This would be REALLY helpful.
We would love to use passman, but we need a "shared vault" for our team. Individual password sharing is not that great here :)

**Comment by chrisHRD** [comment]
Yes, this would be a very necessary feature. Our small team is looking to replace a shared Keepass database with Passman running off a dedicated/isolated Raspberry Pi 3. So far the Nextcloud snap works just fine on the Pi, we just need Passman shared vaults to complete the transition!

**Comment by animalillo** [comment]
Don't you worry, we have not forgotten this feature! It will come, we have had almost no free time lately, busy with life, work and fixing bugs on the extension and the android app.

This feature is on the top of our priority list and we have even give it a little kick  even if you don't see anything yet, took some days off from work to make this feature, but hadn't been enough, I would love to get it finished and released as everybody seems to need this feature, but the way passman is designed difficult adding complex features like this that would look pretty simple to implement on other non client-side encrypted systems.

**Comment by Ninos** [comment]
May you can work in basis of NC13 :-)
https://nextcloud.com/news/

**Comment by Sideboard** [comment]
We are also seeing forward to the new shared vault feature.

In the meantime, we wanted to share some single keys within a user group. The share form suggest that this is possible but is it implemented? Doesn't really warrant a bug report at this point so I'm asking here. Hope that's ok.

**Comment by fredt34** [comment]
We also need to share sets of keys in our team between developers, sysadmins, sales guys: not all of them access all keys. 

Is client-side password encryption such a good idea btw?? https://stackoverflow.com/questions/4121629/password-encryption-at-client-side#4121657

**Comment by fredt34** [comment]
Hmm my previous remark about encryption was stupid, apologies.

**Comment by JurassicTommy** [comment]
This would be amazing and would definitely tip the scales all the way to passman for us to use in our team!

**Comment by rollollooo** [comment]
Hi @brantje ,

first of all thank you and the team for the great work so far. I would also love to see a team sharing and/or sharing of a complete tresor.
Actually we are searching for a webbased password management solution in our team. We already use owncloud in our company and passman is a very interessting alternative to running a password management system in it's own enviroment. But the team/vault sharing feature is must have to use passman in teams or hole companys.
I will talk to the owner of my company to donate some money for this feature. Is there a amount of money you need to get this feature implemented (maybe a kickstarter project)?

regards

**Comment by Braintelligence** [comment]
This missing function is what's keeping me from using Passman yet... it would be awesome if I could just pass vaults over to my clients.

**Comment by xrkolovos** [comment]
this project has started being dead. No one has written anything since summer. The owners are propably busy, and no one has taken the lead. 
We were waiting for this feature since last sprint, so we left passman behind.

**Comment by JurassicTommy** [comment]
Unfortunately we have also moved on from passman and have recently adopted another product. No more need from us.

**Comment by Braintelligence** [comment]
@JurassicTommy Is the other product also open-source?

**Comment by metalcated** [comment]
I would recommend taking this off topic discussion somewhere out of respect
even if this thread is mostly been quiet.

**Comment by marcmanusch** [comment]
Any news ?

**Comment by fir3wall** [comment]
I don't believe this project is maintained anymore ...

**Comment by enoch85** [comment]
This is open source which means people work in their spare time. In this case @brantje is busy with regular work and don't have so much time to spend on this app. With that said, this app is not **not** maintained, it's just that lack of time has made it seem that way.

**Comment by animalillo** [comment]
As @enoch85 said, we still maintain this app, we just don't have time to implement new features.

With that said if there are pull requests we will gladly review and merge them if they fit they improve the project and fit our quality standards.

We hope to get some extra free time and be able to invest it on passman, but currently that's near impossible for us.

**Comment by JurassicTommy** [comment]
As much as I understand and appreciate your position, I don't think you should keep donation links (for specific feature implementation) open whilst the work on it remains seemingly postponed indefinitely. I'm not saying I have donated but it just doesn't seem too fair. But maybe that's just me.

**Comment by wrapper** [comment]
Dropping a + 1 for this feature, unfortunately the system has no use for me without it

**Comment by enoch85** [comment]
Please stop with +1s, Everyone in this issue gets notified. It's much better to hit the like button on the original post instead.

Thank you.

**Comment by KimTheFirst** [comment]
For the reference of anyone who hits this and is desperate; I just got some limited group based sharing going but I'm probably not going to use it. I did it by hacking the credentialMapper and some other files along the lines of:

public function getCredentialsByVaultId($vault_id, $user_id) {
        if(VaultShare::isShared($vault_id,$user_id)){ <<<Look up a table of shares!
            $sql = 'SELECT * FROM `*PREFIX*passman_credentials` ' .
                'WHERE vault_id = ?'; <<Query no longer conditional on user ID being a match!
        }else{
            $sql = 'SELECT * FROM `*PREFIX*passman_credentials` ' .
                'WHERE `user_id` = ? and vault_id = ?';
        }

and so forth. I have no UI for creating the table and it's just a straight up two column table, but hopefully you get the idea. This is literally all you need to do as long as you can figure out how to get data into that table / array / etc. It does not tie in to any other nextcloud features, groups, etc, probably opens up security issues, causes nuclear war, etc. But it does work.

**Comment by LordVan** [comment]
I thought of one workaround (at least for teams): create a seperate user in nextcloud, create a vault and share those passwords within the team. -> by far not ideal, but I'd be more confident in that in a production system than some hack (no offense @KimTheFirst .. you said yourself you probably wouldn't use it either though ;) 

That said .. proper sharing of a whole vault would still be best of course . what is the status of this? @brantje do you think and/or someone else will have time to look at this in the near future?

**Comment by adocampo** [comment]
I would also ask for this, as I'm facing that need on a organization used to LastPass and I've convinced to change to Passman :smiling_imp:

**Comment by RebootfixesAll** [comment]
I would love this feature
let me know if there is any way i can help speed along the process.

**Comment by michaelkarrer81** [comment]
+1 would be a great feature!

**Comment by MrMEEE** [comment]
+1

**Comment by robin-thoni** [comment]
What about storing the vault in a regular file that can then be shared using Nextcloud built-in share feature?
That also means all users would share the same vault password, which sounds acceptable for me.

**Comment by larsziegelmann** [comment]
What's the status of this issue? Is anybody working on it?

**Comment by tibring** [comment]
Could you give us an update of the estimate?

**Comment by mainmachine** [comment]
I had high hopes for this app, but at this point I've switched to bitwarden. I think NC needs to drop passman and integrate with an established password manager - bitwarden is the obvious choice for the license and self-hosting capability.

**Comment by metalcated** [comment]
I also use BitWarden. Great app

**Comment by LokeYourLord** [comment]
So we're in 2023, and yet nothing on this. And all that would be necessary, is to just assign a Vault to a "Group" rather than a single individual, and it would show up for everyone just like Group Shared folders. Shame it hasn't been implemented even after 6 years, would've been a great collaboration tool for NextCloud.

**Comment by simaoafonso-pwt** [comment]
> And all that would be necessary, is to just assign a Vault to a "Group" rather than a single individual, and it would show up for everyone just like Group Shared folders.

If it's so simple, where's your patch?

---

I think integrating to Vaultwarden would be a better use of resources than implementing this, as @mainmachine mentioned before.

**Comment by knixx-fm** [comment]
This is, what a dead repo looks like. The project got stuck on the main feature for a cloud: sharing

**Comment by mainmachine** [comment]
> Yet another comment on this "simple" issue without a link to a Merge Request, wasting the time of the 40 people subscribed to this issue...

Actually there is value in their comment - it's reminded me to *unsubscribe* from this issue. :rofl:

**Comment by knixx-fm** [comment]
Hi!

It's me again.

Thank's for pushing so hard on this issue but we can all take a break now. The issue can be closed. The solution is:
https://github.com/bitwarden](https://github.com/bitwarden)

Thank's to all the developers, that came up with the brilliant idea to spoil their users with kind support and open ears for all those years, on this single issue.

Have good night and tip your waitresses!
